dexcom/fetch/translate.go

```diff
@@ -40,12 +40,31 @@ const MinimumOffsets = (-12 * time.Hour) / OffsetDuration // Minimum time zone o
 const DailyDuration = 24 * time.Hour
 const DailyOffsets = DailyDuration / OffsetDuration
 
-func translateTime(systemTime *dexcom.Time, displayTime *dexcom.Time, datum *dataTypes.Base) {
+// Expectations:
+// - systemTime must not be nil
+// - displayTime can be nil (systemTime used, if so)
+// - datum must not be nil
+func TranslateTime(systemTime *dexcom.Time, displayTime *dexcom.Time, datum *dataTypes.Base) {
```

Comment:
**Not asking for a change here, just hoping you can clarify something for me.**

Is this being exported just so you can call it directly from your test package?

Are you a fan of the separate test package paradigm? I've never understood it. Could you explain the benefits as you see them to me?

*If you said "no" to my first question of the previous paragraph, then you can probably just mark this resolved now and move on.* :)

I'll try to explain my current understanding of the concept. If you have a different understanding, I would be glad to learn from your understanding on the matter.

I understand that the point of having a separately named test package is to enforce that one can only test things via a minimal exported interface. There are two goals of this restriction: 

1. Maintain a minimal exported/public interface.
2. Maintain the ability to refactor the implementation of said interface without breaking tests. 

Does that align with your understanding?

I don't think that works in Go, does it? Go packages make a lot of use of dependency injection, and many of the injected objects are test mocks that specifically check implementation. Take for example the popular [gomock](https://pkg.go.dev/github.com/golang/mock/gomock) package[^1].

If the aim of having a separate package aim is to make it harder to test an interface's implementation (goal 2 above), then it seems like it's failed at that goal; dependency injection is widespread in Go, and there are numerous examples of mock DI objects being used to test implementation.

Moving on to the second goal, as soon as a single function is exported that isn't part of that minimal exported interface, goal 1 above is broken[^2]. That said, there are plenty of cases in our code were function are exported only so they can be tested.

At this point, neither of the stated goals are reached, yet there's significant friction around testing that results from the extra test package. It feels like there's no upside.

So to summarize, I feel like the goals of testing behavior and not implementation, and maintaining a minimal exported interface are good ones, but that using a separate test package doesn't bring a project closer to those goals, in my experience, it likely just makes it harder. For now the best solution might be to aim for disciplined test writing and code reviews to identify tests that are brittle due to testing implementation, and to rewrite them in better ways. 

[1]: I think Gomock is a useful tool, but I believe it's often misued in cases like this.

[2]: This seems to happen a lot. I assume because it's often hard to test an internal function via an external interface. The extra setup and teardown of internal state is a pain. Maybe that's supposed to be an indicator that the internal function should be in its own package? But that doesn't feel very Go like at all to me.

Reply:
*To be clear, not asking for anything to change, just wanting to understand if I'm missing some perspective on how using test packages are in any way useful.*

> I do tend to favor separate test packages so internal implementation details aren't generally tested.

Am I missing something, or does exporting a function solely so it can be tested, exactly contradict your stated reason for having a test package? As I see it a non-exported function is, by definition, "internal implementation details".

> From my perspective, a pragmatic accommodation to balance competing needs.

I'm all for pragmatism. I understand the pain of all that work to test things sometimes. 

What I don't understand, is what need is fulfilled with the test package? 

It can't be the prevention of testing implementation details, as that's not enforced by the test package, and even if it were, it's side-stepped regularly.

> At this point, as platform only uses the separate test package paradigm, I don't think changing it either globally or even for just this package makes sense.

Understood, not trying to ask for any changes, just trying to see if there's some benefit to the practice that I haven't understood.

Reply:
I suppose the root "problem" here is that the translation code is in the same package as the runner. If I had simply created a "dexcom/fetch/translate" package and exported all of the various translation functions from that, then we wouldn't be having this discussion. Yes, I suppose I could do that after-the-fact, but that seems like more work than benefit.

Reply:
Thanks for taking the time to answer my questions, I appreciate it.